### PR TITLE
Throw error on null/empty Id created

### PR DESCRIPTION
**Pull request by PatriceKammognie**

### CODE REVIEW

File: basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/core/ConceptDescriptionRepositorySuite.java [hunk 1]
```diff
@@ -259,7 +260,7 @@ private void assertConceptDescriptionsAreContained(Collection<ConceptDescription
 		assertTrue(actualConceptDescriptions.containsAll(expectedConceptDescriptions));
 	}
 
-	private ConceptDescription createDummyConceptDescription(String id) {
+	protected ConceptDescription createDummyConceptDescription(String id) {
```
Comment (mdanish98) [inline basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/core/ConceptDescriptionRepositorySuite.java:263]: When the test cases are moved in this class (as suggested above) then this change is not required.

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java [hunk 2]
```diff
@@ -338,6 +341,11 @@ private void throwIfSubmodelDoesNotExist(String id) {
 		if (!submodelServices.containsKey(id))
 			throw new ElementDoesNotExistException(id);
 	}
+	
+	private void throwIfSubmodelIdEmptyOrNull(String id) {
+		if(id == null || id.trim().isEmpty())
```
Comment (mdanish98) [inline basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java:346]: I think isBlank() is appropriate here.

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java [hunk 3]
```diff
@@ -173,9 +174,11 @@ public void updateSubmodel(String id, Submodel submodel) throws ElementDoesNotEx
 	}
 
 	@Override
-	public void createSubmodel(Submodel submodel) throws CollidingIdentifierException {
+	public void createSubmodel(Submodel submodel) throws CollidingIdentifierException, MissingIdentifierException {
+		throwIfSubmodelIdEmptyOrNull(submodel.getId());
 		throwIfSubmodelExists(submodel.getId());
-
+		
+
```
Comment (mdanish98) [inline basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java:181]: Please remove this unnecessary line.

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java [hunk 4]
```diff
@@ -173,9 +174,11 @@ public void updateSubmodel(String id, Submodel submodel) throws ElementDoesNotEx
 	}
 
 	@Override
-	public void createSubmodel(Submodel submodel) throws CollidingIdentifierException {
+	public void createSubmodel(Submodel submodel) throws CollidingIdentifierException, MissingIdentifierException {
+		throwIfSubmodelIdEmptyOrNull(submodel.getId());
 		throwIfSubmodelExists(submodel.getId());
```
Comment (mdanish98) [inline basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java:179]: Please consider adding a vertical space above this.

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/TestInMemorySubmodelRepository.java
Comment (mdanish98) [inline basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/TestInMemorySubmodelRepository.java:1]: The test cases added here should be in SubmodelRepositorySuite.
Please move these cases to SubmodelRepositorySuite and reset this test class.

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java
Comment (mdanish98) [inline basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java:1]: InMemorySMRepo has a constructor that takes a collection of Submodels. Here, each Submodel should be checked if their IDs are either null or empty, similar to the already existing check throwIfHasCollidingIds(submodels)

File: basyx.submodelrepository/basyx.submodelrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/SubmodelRepository.java [hunk 5]
```diff
@@ -79,7 +80,7 @@ public interface SubmodelRepository {
 	 * @param submodel
 	 * @throws CollidingIdentifierException
 	 */
-	public void createSubmodel(Submodel submodel) throws CollidingIdentifierException;
+	public void createSubmodel(Submodel submodel) throws CollidingIdentifierException,MissingIdentifierException;
```
Comment (mdanish98) [inline basyx.submodelrepository/basyx.submodelrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/SubmodelRepository.java:83]: Please use BaSyx formatter which is available [here](https://wiki.eclipse.org/File:BaSyx_Formatting.zip).

File: basyx.submodelrepository/basyx.submodelrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/core/SubmodelRepositorySuite.java [hunk 6]
```diff
@@ -523,7 +523,7 @@ private SubmodelElement getExpectedSubmodelElement() {
 				.get();
 	}
 
-	private Submodel buildDummySubmodel(String id) {
+	protected Submodel buildDummySubmodel(String id) {
```
Comment (mdanish98) [inline basyx.submodelrepository/basyx.submodelrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/core/SubmodelRepositorySuite.java:526]: When the test cases are moved in this class (as suggested above) then this change is not required.

**Review by mdanish98 — CHANGES_REQUESTED** [review]
Some general remarks are as below:
- There is no implementation coverage for MongoDB, please add this in each repo.
- HTTP related tests are missing, please add these in each repo

File: basyx.aasrepository/basyx.aasrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/AasRepository.java [hunk 7]
```diff
@@ -64,7 +65,7 @@ public interface AasRepository {
 	 * @param aas
 	 *            the AAS to be created
 	 */
-	public void createAas(AssetAdministrationShell aas) throws CollidingIdentifierException;
+	public void createAas(AssetAdministrationShell aas) throws CollidingIdentifierException, MissingIdentifierException ;
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.aasrepository/basyx.aasrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/AasRepository.java:68]: done

File: basyx.aasrepository/basyx.aasrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/AasRepository.java [hunk 8]
```diff
@@ -64,7 +65,7 @@ public interface AasRepository {
 	 * @param aas
 	 *            the AAS to be created
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.aasrepository/basyx.aasrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/AasRepository.java:66]: done

File: basyx.aasrepository/basyx.aasrepository-backend/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/backend/CrudAasRepository.java [hunk 9]
```diff
@@ -88,7 +89,8 @@ public AssetAdministrationShell getAas(String aasId) throws ElementDoesNotExistE
 	}
 
 	@Override
-	public void createAas(AssetAdministrationShell aas) throws CollidingIdentifierException {
+	public void createAas(AssetAdministrationShell aas) throws CollidingIdentifierException, MissingIdentifierException {
+		throwIfAasIdEmptyOrNull(aas.getId());
 		throwIfAasExists(aas);
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.aasrepository/basyx.aasrepository-backend/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/backend/CrudAasRepository.java:94]: done

File: basyx.aasrepository/basyx.aasrepository-backend/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/backend/CrudAasRepository.java [hunk 10]
```diff
@@ -164,6 +166,11 @@ private void throwIfAasExists(AssetAdministrationShell aas) {
 		if (aasBackend.existsById(aas.getId()))
 			throw new CollidingIdentifierException();
 	}
+	
+	private void throwIfAasIdEmptyOrNull(String id) {
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.aasrepository/basyx.aasrepository-backend/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/backend/CrudAasRepository.java:170]: done

File: basyx.aasrepository/basyx.aasrepository-backend/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/backend/CrudAasRepository.java [hunk 11]
```diff
@@ -164,6 +166,11 @@ private void throwIfAasExists(AssetAdministrationShell aas) {
...
+		if(id == null || id.trim().isEmpty())
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.aasrepository/basyx.aasrepository-backend/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/backend/CrudAasRepository.java:171]: done

File: basyx.aasrepository/basyx.aasrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/AasRepositorySuite.java [hunk 12]
```diff
@@ -299,6 +314,11 @@ private AssetInformation createDummyAssetInformation() {
 				.globalAssetID("assetIDTestKey")
 				.build();
 	}
+	
+	public static AssetAdministrationShell createDummyAssetAdministrationShell(String AasId) {
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.aasrepository/basyx.aasrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/AasRepositorySuite.java:318]: done

File: basyx.common/basyx.core/src/main/java/org/eclipse/digitaltwin/basyx/core/exceptions/MissingIdentifierException.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ * 
+ * SPDX-License-Identifier: MIT
+ ******************************************************************************/
+
+ package org.eclipse.digitaltwin.basyx.core.exceptions;
+
+ /**
+  * Indicates that an entity has no identifier or an Identifier with the value NULL
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.common/basyx.core/src/main/java/org/eclipse/digitaltwin/basyx/core/exceptions/MissingIdentifierException.java:29]: done

File: basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/InMemoryConceptDescriptionRepository.java [hunk 14]
```diff
@@ -164,7 +165,8 @@ public void updateConceptDescription(String conceptDescriptionId, ConceptDescrip
 	}
 
 	@Override
-	public void createConceptDescription(ConceptDescription conceptDescription) throws CollidingIdentifierException {
+	public void createConceptDescription(ConceptDescription conceptDescription) throws CollidingIdentifierException,MissingIdentifierException {
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/InMemoryConceptDescriptionRepository.java:168]: done

File: basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/InMemoryConceptDescriptionRepository.java [hunk 15]
```diff
@@ -204,6 +206,11 @@ private void throwIfConceptDescriptionExists(String id) {
 		if (conceptDescriptions.containsKey(id))
 			throw new CollidingIdentifierException(id);
 	}
+	
+	private void throwIfConceptDescriptionIdEmptyOrNull(String id) {
+		if(id == null || id.trim().isEmpty())
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/InMemoryConceptDescriptionRepository.java:211]: done

File: basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/ConceptDescriptionRepository.java [hunk 16]
```diff
@@ -100,7 +101,7 @@ public interface ConceptDescriptionRepository {
 	 * @param conceptDescription
 	 * @throws CollidingIdentifierException
 	 */
-	public void createConceptDescription(ConceptDescription conceptDescription) throws CollidingIdentifierException;
+	public void createConceptDescription(ConceptDescription conceptDescription) throws CollidingIdentifierException,MissingIdentifierException;
```
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/ConceptDescriptionRepository.java:104]: done

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java (hunk 2 above)
Comment (PatriceKammognie) [reply basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java:346]: done

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java (hunk 3 above)
Comment (PatriceKammognie) [reply basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java:181]: done

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java (hunk 4 above)
Comment (PatriceKammognie) [reply basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java:179]: done

File: basyx.submodelrepository/basyx.submodelrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/SubmodelRepository.java (hunk 5 above)
Comment (PatriceKammognie) [reply basyx.submodelrepository/basyx.submodelrepository-core/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/SubmodelRepository.java:83]: done

**Comment by FrankSchnicke** [comment]
Closes https://github.com/eclipse-basyx/basyx-java-server-sdk/issues/68

File: basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/InMemoryConceptDescriptionRepository.java
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/InMemoryConceptDescriptionRepository.java:1]: done

File: basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/test/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/TestInMemoryConceptDescriptionRepository.java
[the remark this answers is not in the document]
Comment (PatriceKammognie) [reply basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/test/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/TestInMemoryConceptDescriptionRepository.java:1]: done

File: basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/core/ConceptDescriptionRepositorySuite.java (hunk 1 above)
Comment (PatriceKammognie) [reply basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/core/ConceptDescriptionRepositorySuite.java:263]: done

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/TestInMemorySubmodelRepository.java
Comment (PatriceKammognie) [reply basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/TestInMemorySubmodelRepository.java:1]: done

File: basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java
Comment (PatriceKammognie) [reply basyx.submodelrepository/basyx.submodelrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/InMemorySubmodelRepository.java:1]: done

File: basyx.submodelrepository/basyx.submodelrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/core/SubmodelRepositorySuite.java (hunk 6 above)
Comment (PatriceKammognie) [reply basyx.submodelrepository/basyx.submodelrepository-core/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/core/SubmodelRepositorySuite.java:526]: done

File: basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/InMemoryConceptDescriptionRepository.java [hunk 17]
```diff
@@ -196,35 +188,40 @@ private static void assertIdUniqueness(Collection<ConceptDescription> conceptDes
...
+		return conceptDescriptions.stream().collect(Collectors.toMap(ConceptDescription::getId, conceptDescription -> conceptDescription));
 	}
 
 	private void throwIfConceptDescriptionExists(String id) {
 		if (conceptDescriptions.containsKey(id))
 			throw new CollidingIdentifierException(id);
 	}
 
+	private void throwIfMissingId(Collection<ConceptDescription> conceptDescriptionsToCheck) {
+		boolean isMissingId = conceptDescriptionsToCheck.stream().map(ConceptDescription::getId).anyMatch(id -> id == null || id.isBlank());
```
Comment (mdanish98) [inline basyx.conceptdescriptionrepository/basyx.conceptdescriptionrepository-backend-inmemory/src/main/java/org/eclipse/digitaltwin/basyx/conceptdescriptionrepository/InMemoryConceptDescriptionRepository.java:200]: Please verify the working.

**Pull request closed without merging**